src/io/CavityWriter.cpp

```diff
@@ -10,96 +10,248 @@
+        global_log->info() << "[CavityWriter] Component: " << componentID << endl;
+        CavityEnsemble cav = CavityEnsemble();
+        _mcav[componentID] = cav;
+    }
+
+    // go back to root
+    xmlconfig.changecurrentnode(oldpath);
 }
 
-void CavityWriter::init(ParticleContainer * /*particleContainer*/, DomainDecompBase * /*domainDecomp*/,
-						Domain * /*domain*/) {}
+void CavityWriter::init(ParticleContainer * particleContainer, DomainDecompBase * domainDecomp,
+						Domain * domain) {
+
+    // set initial values for CavityEnsemble
+
+	double Tcur = domain->getGlobalCurrentTemperature();
+
+	/* FIXME: target temperature from thermostat ID 0 or 1? */
+	double Ttar = domain->severalThermostats() ? domain->getTargetTemperature(1)
+												 : domain->getTargetTemperature(0);
+	if ((Tcur < 0.85 * Ttar) || (Tcur > 1.15 * Ttar))
+		Tcur = Ttar;
+
+	map<unsigned, CavityEnsemble>::iterator ceit;
+	for (ceit = _mcav.begin(); ceit != _mcav.end(); ceit++) {
+
+        // setup
+        ceit->second.setSystem(domain->getGlobalLength(0), domain->getGlobalLength(1), domain->getGlobalLength(2), _maxNeighbors, _radius);
+        int rank = domainDecomp->getRank();
+        double min[3], max[3];
+        domainDecomp->getBoundingBoxMinMax(domain, min, max);
+        ceit->second.setSubdomain(rank, min[0], max[0], min[1], max[1], min[2], max[2]);
+        // optional here: setControlVolume
+		ceit->second.submitTemperature(Tcur);
+		int cID = ceit->first;
+		Component* c = global_simulation->getEnsemble()->getComponent(cID);
+		global_log->info() << "[Cavity Writer] init: " << cID << endl;
+		// TODO: Nx, Ny, Nz via XML params
+        ceit->second.init(c, _Nx, _Ny, _Nz);
+        global_log->info() << "[Cavity Writer] init done: " << cID << endl;
+    }
+	//Simulation::exit(999);
+}
+
+void CavityWriter::beforeEventNewTimestep(
+		ParticleContainer* particleContainer, DomainDecompBase* domainDecomp,
+		unsigned long simstep
+) {
+	if (simstep >= global_simulation->getInitStatistics() && simstep % _writeFrequency == 0) {
+		map<unsigned, CavityEnsemble>::iterator ceit;
+		for(ceit = this->_mcav.begin(); ceit != this->_mcav.end(); ceit++) {
+			//if (!((simstep + 2 * ceit->first + 3) % ceit->second.getInterval())) {
+			//	ceit->second.preprocessStep();
+			//}
+            ceit->second.preprocessStep();
+		}
+	}
+}
+
+void CavityWriter::afterForces(
+		ParticleContainer* particleContainer, DomainDecompBase* domainDecomp,
+		unsigned long simstep
+) {
+
+	if(simstep >= global_simulation->getInitStatistics() && simstep % _writeFrequency == 0) {
+		map<unsigned, CavityEnsemble>::iterator ceit;
+		for(ceit = this->_mcav.begin(); ceit != this->_mcav.end(); ceit++) {
+
+			unsigned cavityComponentID = ceit->first;
+			CavityEnsemble & cavEns = ceit->second;
+
+			//TODO: cavEns calculate Interval / max Neighbors, via XML
+
+			/*
+			        if (!((simstep + 2 * cavityComponentID + 3) % cavEns.getInterval())) {
```

Comment:
Why is this part commented out now?

Reply:
Previously, CavityEnsemble would have its own interval check as it was called directly via Simulation loop. Now, CavityEnsemble is only called when CavityWriter hits a writeStep, so there is no need for CavityEnsemble itself to check simsteps anymore.